runner/check.go

```diff
@@ -0,0 +1,120 @@
+package runner
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os/exec"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// Check is a basic structure that describes DC/OS check.
+type Check struct {
+	// Cmd is a path to executable (script or binary) and arguments
+	Cmd []string `json:"cmd"`
+
+	// Description provides a basic check description.
+	Description string `json:"description"`
+
+	// Timeout is defines a custom script timeout.
+	Timeout string `json:"timeout"`
+
+	// Roles is a list of DC/OS roles (e.g. master, agent). Node must be one of roles
+	// to execute a check.
+	Roles []string `json:"roles"`
+}
+
+// Run executes the given check.
+func (c *Check) Run(ctx context.Context, role string) ([]byte, error) {
+	var (
+		// define an empty cancel function
+		cancel context.CancelFunc = func() {}
+
+		arg []string
+	)
+
+	if !c.verifyRole(role) {
+		return nil, errors.Errorf("check can be executed on a node with the following roles %s. Current role %s", c.Roles, role)
+	}
+
+	if len(c.Cmd) == 0 {
+		return nil, errors.New("unable to execute a command with empty Cmd field")
+	}
+
+	if ctx == nil {
+		// default to 10 seconds timeout.
+		ctx, cancel = context.WithTimeout(context.Background(), time.Second*10)
+	}
+
+	if c.Timeout != "" {
+		timeout, err := time.ParseDuration(c.Timeout)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid timeout string %s", timeout)
+		}
+
+		ctx, cancel = context.WithTimeout(ctx, timeout)
```

Comment:
If ctx is nil above, we create a new one with a 10s timeout.  What happens when you create another context based on that context (with the default 10s timeout) using another timeout that can be larger or smaller?